File: javascript/438_-_Find_All_Anagrams_in_a_String.js
/**
 * 438. Find All Anagrams in a String - Medium
 * Given two strings s and p, return an array of all the start indices of p's anagrams in s.
 * You may return the answer in any order.
 * An Anagram is a word or phrase formed by rearranging the letters of a different word or
 * phrase, typically using all the original letters exactly once.
 */

/**
 * @param {string} s
 * @param {string} p
 * @return {number[]}
 */
var findAnagrams = function(s, p) {
    const freq = new Map();
    const ans = [];

    for (const c of p) {
        if (freq.has(c)) {
            freq.set(c, freq.get(c) + 1);
        } else {
            freq.set(c, 1);
        }
    }

    let begin = 0;
    let end = 0;
    let counter = freq.size;

    while (end < s.length) {
        const endChar = s.charAt(end);

        if (freq.has(endChar)) {
            freq.set(endChar, freq.get(endChar) - 1);

            if (freq.get(endChar) === 0) {
                --counter;
            }
        }

        ++end;

        // limit the length of our window to the length of p
        if (end - begin === p.length) {
            if (counter === 0) {
                ans.push(begin);
            }

            const beginChar = s.charAt(begin);

            if (freq.has(beginChar)) {
                freq.set(beginChar, freq.get(beginChar) + 1);

                if (freq.get(beginChar) === 1) {
                    ++counter;
                }
            }

            ++begin;
        }
    }

    return ans;
};

File: csharp/438_-_Find_All_Anagrams_in_a_String.cs
/**
 * 438. Find All Anagrams in a String - Medium
 * Given two strings s and p, return an array of all the start indices of p's anagrams in s.
 * You may return the answer in any order.
 * An Anagram is a word or phrase formed by rearranging the letters of a different word or
 * phrase, typically using all the original letters exactly once.
 */

public class Solution {
    public IList<int> FindAnagrams(string s, string p) {
        Dictionary<char,int> freq = new Dictionary<char,int>();
        IList<int> ans = new List<int>();
        
        foreach (char c in p) {
            if (freq.ContainsKey(c)) {
                freq[c] += 1;
            } else {
                freq[c] = 1;
            }
        }
        
        int begin = 0;
        int end = 0;
        int counter = freq.Count;
        
        while (end < s.Length) {
            char endChar = s[end];
            
            if (freq.ContainsKey(endChar)) {
                freq[endChar] -= 1;
                
                if (freq[endChar] == 0) {
                    --counter;
                }
            }
            
            ++end;
            
            // limit the length of our window to the length of p
            if (end - begin == p.Length) {
                if (counter == 0) {
                    ans.Add(begin);
                }
                
                char beginChar = s[begin];
                
                if (freq.ContainsKey(beginChar)) {
                    freq[beginChar] += 1;
                    
                    if (freq[beginChar] == 1) {
                        ++counter;
                    }
                }
                
                ++begin;
            }
        }
        
        return ans;
    }
}